File: SubAgenticAI.Core/SubAgent.cs
namespace SubAgenticAI.Core;

/// <summary>
/// Implementation of a SubAgent - specialized, mission-scoped cognitive unit with strict isolation
/// </summary>
public class SubAgent : ISubAgent
{
    private readonly Dictionary<string, object> _localState;
    private readonly List<string> _activityLog;
    
    public string Id { get; }
    public string Domain { get; }
    public MissionStatus Status { get; private set; }
    public TrustProfile TrustProfile { get; }
    
    /// <summary>
    /// Creates a new SubAgent
    /// </summary>
    public SubAgent(string id, string domain, TrustProfile trustProfile)
    {
        Id = id ?? throw new ArgumentNullException(nameof(id));
        Domain = domain ?? throw new ArgumentNullException(nameof(domain));
        TrustProfile = trustProfile ?? throw new ArgumentNullException(nameof(trustProfile));
        
        _localState = new Dictionary<string, object>();
        _activityLog = new List<string>();
        Status = MissionStatus.Created;
        
        LogActivity($"SubAgent created: {Id} for domain: {Domain}");
    }
    
    public Task StartMissionAsync()
    {
        if (Status != MissionStatus.Created)
        {
            throw new InvalidOperationException($"Cannot start mission from status: {Status}");
        }
        
        Status = MissionStatus.Active;
        LogActivity("Mission started");
        return Task.CompletedTask;
    }
    
    public async Task<string> ProcessTaskAsync(string task)
    {
        if (Status != MissionStatus.Active)
        {
            throw new InvalidOperationException($"Cannot process task in status: {Status}");
        }
        
        LogActivity($"Processing task: {task}");
        
        // This is a placeholder for actual task processing logic
        // In a real implementation, this would involve domain-specific processing
        await Task.Delay(10); // Simulate processing
        
        var result = $"Processed: {task} in domain: {Domain}";
        LogActivity($"Task completed: {result}");
        
        return result;
    }
    
    public async Task<bool> ReceiveInformationAsync(string senderId, string information)
    {
        if (Status != MissionStatus.Active)
        {
            LogActivity($"Rejected inbound information - status: {Status}");
            return false;
        }
        
        // Validate inbound information using trust profile
        var isValid = await TrustProfile.ValidateInboundAsync(senderId, information);
        
        if (isValid)
        {
            _localState[$"received_{DateTime.UtcNow.Ticks}"] = information;
            LogActivity($"Accepted information from: {senderId}");
            return true;
        }
        
        LogActivity($"Rejected information from: {senderId} - trust validation failed");
        return false;
    }
    
    public Task<bool> SendInformationAsync(string recipientId, string information)
    {
        if (Status != MissionStatus.Active)
        {
            LogActivity($"Cannot send information - status: {Status}");
            return Task.FromResult(false);
        }
        
        // Security level check for outbound information
        // In a real implementation, this would check against recipient's clearance
        LogActivity($"Sending information to: {recipientId} (SL: {TrustProfile.SecurityLevel})");
        return Task.FromResult(true);
    }
    
    public Task CompleteMissionAsync()
    {
        if (Status != MissionStatus.Active && Status != MissionStatus.Paused)
        {
            throw new InvalidOperationException($"Cannot complete mission from status: {Status}");
        }
        
        Status = MissionStatus.Completed;
        LogActivity("Mission completed");
        return Task.CompletedTask;
    }
    
    public Task RetireAsync()
    {
        if (Status == MissionStatus.Retired)
        {
            return Task.CompletedTask;
        }
        
        Status = MissionStatus.Retired;
        LogActivity("SubAgent retired");
        
        // Clear local state for garbage collection
        _localState.Clear();
        
        return Task.CompletedTask;
    }
    
    /// <summary>
    /// Gets the activity log for auditing purposes
    /// </summary>
    public IReadOnlyList<string> GetActivityLog() => _activityLog.AsReadOnly();
    
    /// <summary>
    /// Gets a snapshot of local state (for debugging/auditing)
    /// </summary>
    public IReadOnlyDictionary<string, object> GetStateSnapshot() => 
        new Dictionary<string, object>(_localState);
    
    private void LogActivity(string activity)
    {
        var logEntry = $"[{DateTime.UtcNow:yyyy-MM-dd HH:mm:ss.fff}] {activity}";
        _activityLog.Add(logEntry);
    }
}

File: subagentic_ai_py/subagentic_ai/sub_agent.py
"""SubAgent interface and implementation."""

from abc import ABC, abstractmethod
from datetime import datetime
from typing import Dict, List, Any

from .mission_status import MissionStatus
from .trust_profile import TrustProfile


class ISubAgent(ABC):
    """Interface for a SubAgent (SA) - a specialized, mission-scoped cognitive unit."""
    
    @property
    @abstractmethod
    def id(self) -> str:
        """Unique identifier for this SubAgent."""
        pass
    
    @property
    @abstractmethod
    def domain(self) -> str:
        """Domain or mission focus of this SubAgent."""
        pass
    
    @property
    @abstractmethod
    def status(self) -> MissionStatus:
        """Current mission status."""
        pass
    
    @property
    @abstractmethod
    def trust_profile(self) -> TrustProfile:
        """Trust profile assigned by the FLA."""
        pass
    
    @abstractmethod
    async def start_mission(self) -> None:
        """Start the SubAgent's mission."""
        pass
    
    @abstractmethod
    async def process_task(self, task: str) -> str:
        """Process a task within the SubAgent's domain."""
        pass
    
    @abstractmethod
    async def receive_information(self, sender_id: str, information: str) -> bool:
        """Receive information from external source (validated by trust profile)."""
        pass
    
    @abstractmethod
    async def send_information(self, recipient_id: str, information: str) -> bool:
        """Send information outbound (governed by security level)."""
        pass
    
    @abstractmethod
    async def complete_mission(self) -> None:
        """Complete the mission."""
        pass
    
    @abstractmethod
    async def retire(self) -> None:
        """Retire/dispose the SubAgent."""
        pass


class SubAgent(ISubAgent):
    """Implementation of a SubAgent - specialized, mission-scoped cognitive unit with strict isolation."""
    
    def __init__(self, agent_id: str, domain: str, trust_profile: TrustProfile):
        """
        Create a new SubAgent.
        
        Args:
            agent_id: Unique identifier
            domain: Domain or mission focus
            trust_profile: Trust profile assigned by FLA
        """
        self._id = agent_id
        self._domain = domain
        self._trust_profile = trust_profile
        self._status = MissionStatus.CREATED
        self._local_state: Dict[str, Any] = {}
        self._activity_log: List[str] = []
        
        self._log_activity(f"SubAgent created: {agent_id} for domain: {domain}")
    
    @property
    def id(self) -> str:
        return self._id
    
    @property
    def domain(self) -> str:
        return self._domain
    
    @property
    def status(self) -> MissionStatus:
        return self._status
    
    @property
    def trust_profile(self) -> TrustProfile:
        return self._trust_profile
    
    async def start_mission(self) -> None:
        """Start the mission."""
        if self._status != MissionStatus.CREATED:
            raise RuntimeError(f"Cannot start mission from status: {self._status}")
        
        self._status = MissionStatus.ACTIVE
        self._log_activity("Mission started")
    
    async def process_task(self, task: str) -> str:
        """Process a task."""
        if self._status != MissionStatus.ACTIVE:
            raise RuntimeError(f"Cannot process task in status: {self._status}")
        
        self._log_activity(f"Processing task: {task}")
        
        # Placeholder for actual task processing logic
        # In a real implementation, this would involve domain-specific processing
        result = f"Processed: {task} in domain: {self._domain}"
        self._log_activity(f"Task completed: {result}")
        
        return result
    
    async def receive_information(self, sender_id: str, information: str) -> bool:
        """Receive information with trust validation."""
        if self._status != MissionStatus.ACTIVE:
            self._log_activity(f"Rejected inbound information - status: {self._status}")
            return False
        
        # Validate inbound information using trust profile
        is_valid = await self._trust_profile.validate_inbound(sender_id, information)
        
        if is_valid:
            self._local_state[f"received_{datetime.utcnow().timestamp()}"] = information
            self._log_activity(f"Accepted information from: {sender_id}")
            return True
        
        self._log_activity(f"Rejected information from: {sender_id} - trust validation failed")
        return False
    
    async def send_information(self, recipient_id: str, information: str) -> bool:
        """Send information outbound."""
        if self._status != MissionStatus.ACTIVE:
            self._log_activity(f"Cannot send information - status: {self._status}")
            return False
        
        # Security level check for outbound information
        self._log_activity(f"Sending information to: {recipient_id} (SL: {self._trust_profile.security_level})")
        return True
    
    async def complete_mission(self) -> None:
        """Complete the mission."""
        if self._status not in (MissionStatus.ACTIVE, MissionStatus.PAUSED):
            raise RuntimeError(f"Cannot complete mission from status: {self._status}")
        
        self._status = MissionStatus.COMPLETED
        self._log_activity("Mission completed")
    
    async def retire(self) -> None:
        """Retire the SubAgent."""
        if self._status == MissionStatus.RETIRED:
            return
        
        self._status = MissionStatus.RETIRED
        self._log_activity("SubAgent retired")
        
        # Clear local state for garbage collection
        self._local_state.clear()
    
    def get_activity_log(self) -> List[str]:
        """Get the activity log for auditing purposes."""
        return self._activity_log.copy()
    
    def get_state_snapshot(self) -> Dict[str, Any]:
        """Get a snapshot of local state (for debugging/auditing)."""
        return self._local_state.copy()
    
    def _log_activity(self, activity: str) -> None:
        """Log an activity."""
        log_entry = f"[{datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S.%f')[:-3]}] {activity}"
        self._activity_log.append(log_entry)
